hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/s3/bucket/S3BucketCreateRequest.java

```diff
@@ -0,0 +1,350 @@
+      if (omMetadataManager.getS3Table().isExist(s3BucketName)) {
+        throw new OMException("S3Bucket " + s3BucketName + " already exists",
+            OMException.ResultCodes.S3_BUCKET_ALREADY_EXISTS);
+      }
+
+      try {
+        acquiredVolumeLock =
+            omMetadataManager.getLock().acquireLock(VOLUME_LOCK, volumeName);
+        acquiredUserLock = omMetadataManager.getLock().acquireLock(USER_LOCK,
+            userName);
+        // Check if volume exists, if it does not exist create
+        // ozone volume.
+        String volumeKey = omMetadataManager.getVolumeKey(volumeName);
+        if (!omMetadataManager.getVolumeTable().isExist(volumeKey)) {
+          omVolumeArgs = createOmVolumeArgs(volumeName, userName,
+              s3CreateBucketRequest.getS3CreateVolumeInfo()
+                  .getCreationTime());
+          volumeList = omMetadataManager.getUserTable().get(
+              omMetadataManager.getUserKey(userName));
+          volumeList = addVolumeToOwnerList(volumeList,
+              volumeName, userName, ozoneManager.getMaxUserVolumeCount());
+          createVolume(omMetadataManager, omVolumeArgs, volumeList, volumeKey,
+              omMetadataManager.getUserKey(userName), transactionLogIndex);
+          volumeCreated = true;
+        }
+      } finally {
+        if (acquiredUserLock) {
+          omMetadataManager.getLock().releaseLock(USER_LOCK, userName);
+        }
+        if (acquiredVolumeLock) {
+          omMetadataManager.getLock().releaseLock(VOLUME_LOCK, volumeName);
+        }
+      }
+
+      // check if ozone bucket exists, if it does not exist create ozone
+      // bucket
+      boolean acquireBucketLock = false;
+      try {
+        acquireBucketLock =
+            omMetadataManager.getLock().acquireLock(BUCKET_LOCK, volumeName,
+                s3BucketName);
+        String bucketKey = omMetadataManager.getBucketKey(volumeName,
+            s3BucketName);
+        if (!omMetadataManager.getBucketTable().isExist(bucketKey)) {
+          omBucketInfo = createOmBucketInfo(volumeName, s3BucketName,
+              s3CreateBucketRequest.getS3CreateVolumeInfo()
+                  .getCreationTime());
+          // Add to bucket table cache.
+          omMetadataManager.getBucketTable().addCacheEntry(
+              new CacheKey<>(bucketKey),
+              new CacheValue<>(Optional.of(omBucketInfo), transactionLogIndex));
+          bucketCreated = true;
+        } else {
+          // This can happen when a ozone bucket exists already in the
+          // volume, but this is not a s3 bucket.
+          throw new OMException("Bucket " + s3BucketName + " already exists",
+              OMException.ResultCodes.BUCKET_ALREADY_EXISTS);
+        }
+      } finally {
+        if (acquireBucketLock) {
+          omMetadataManager.getLock().releaseLock(BUCKET_LOCK, volumeName,
+              s3BucketName);
+        }
+      }
+
+      // Now finally add it to s3 table cache.
```

Comment:
IIUC the same bucket is cached twice, once in bucket table and once in S3 table.

Reply:
In S3Table we cache S3Bucketname -> OzonevolumeName/OzoneBucketName. In S3 world buckets are unique, to acheive this, we have this S3Table.
In Bucket Table we cache BucketName to BucketInfo, not this mapping.